src/main/scala/zio/package.scala

```diff
@@ -341,43 +341,20 @@ package object zmx extends MetricsDataModel with MetricsConfigDataModel {
       ): ZIO[Clock, Throwable, Fiber.Runtime[Throwable, Nothing]]          =
         collect(f).forever.forkDaemon <& sendIfNotEmpty(f).repeat(everyNSec).forkDaemon
 
-      private[zio] val unsafeClient                            = UDPClientUnsafe(config.host.getOrElse("localhost"), config.port.getOrElse(8125))
-      private[zio] val udpUnsafe: List[Metric[_]] => List[Int] = metrics =>
+      private[zio] val unsafeClient                             = UDPClientUnsafe(config.host.getOrElse("localhost"), config.port.getOrElse(8125))
+      private[zio] val udpUnsafe: List[Metric[_]] => List[Long] = metrics =>
         for {
-          flt <- sample(metrics).map(Encoder.encode).toList
-        } yield unsafeClient.send(flt)
-
-      def listenUnsafe(): (ScheduledFuture[_], ScheduledFuture[_]) = listenUnsafe(udpUnsafe)
-      def listenUnsafe(
-        f: List[Metric[_]] => List[Int]
-      ): (ScheduledFuture[_], ScheduledFuture[_]) = {
-        val fixedExecutor = new ScheduledThreadPoolExecutor(2)
-        val collectTask   = new Runnable {
-          def run() =
-            while (true)
-              ring.poll(Metric.Zero) match {
-                case Metric.Zero => ()
-                case m @ _       =>
-                  val l = aggregator.updateAndGet(_ :+ m)
-                  if (l.size == config.bufferSize) {
-                    f(aggregator.getAndUpdate(_ => List.empty[Metric[_]]))
-                  }
-              }
-        }
-
-        val timeoutTask = new Runnable {
-          def run() = {
-            if (!aggregator.get().isEmpty) {
-              f(aggregator.getAndUpdate(_ => List.empty[Metric[_]]))
-            }
-            ()
-          }
-        }
+          flt <- sample(metrics).map(Encoder.encode)
+        } yield unsafeClient.send(flt).toLong
 
-        val rateHook    = fixedExecutor.scheduleAtFixedRate(timeoutTask, 5, config.timeout.toMillis, TimeUnit.MILLISECONDS)
-        val collectHook = fixedExecutor.schedule(collectTask, 500, TimeUnit.MILLISECONDS)
-        (rateHook, collectHook)
-      }
+      def listenUnsafe(): CancelableFuture[Nothing]                                 = listenUnsafe(udpUnsafe)
+      def listenUnsafe(f: List[Metric[_]] => List[Long]): CancelableFuture[Nothing] =
+        Runtime.default.unsafeRunToFuture(
+          for {
+            fiber <- listen(metrics => IO(f(metrics)).refineOrDie { case e: Exception => e })
+            value <- UIO.never.ensuring(fiber.interrupt)
```

Comment:
I think you can do something like `listen(??? ) <* ZIO.never).ensuring(_.interrupt)`. We don't have to do as part of this PR but I think if we make some changes to the `listen` method even this won't be necessary. I think we are forking a fiber there with `effect.forkDaemon` and instead we should do `ZIO.transplant(graft => graft(effect).fork)`. That will make sure that the fiber doesn't get interrupted too early but that the forked fiber is automatically terminated when the main fiber terminates.